source/common/stream_info/utility.cc

```diff
@@ -134,5 +140,240 @@ Utility::formatDownstreamAddressJustPort(const Network::Address::Instance& addre
+  case ProxyStatusError::DestinationUnavailable:
+    return DESTINATION_UNAVAILABLE;
+  case ProxyStatusError::DestinationIpProhibited:
+    return DESTINATION_IP_PROHIBITED;
+  case ProxyStatusError::DestinationIpUnroutable:
+    return DESTINATION_IP_UNROUTABLE;
+  case ProxyStatusError::ConnectionRefused:
+    return CONNECTION_REFUSED;
+  case ProxyStatusError::ConnectionTerminated:
+    return CONNECTION_TERMINATED;
+  case ProxyStatusError::ConnectionTimeout:
+    return CONNECTION_TIMEOUT;
+  case ProxyStatusError::ConnectionReadTimeout:
+    return CONNECTION_READ_TIMEOUT;
+  case ProxyStatusError::ConnectionWriteTimeout:
+    return CONNECTION_WRITE_TIMEOUT;
+  case ProxyStatusError::ConnectionLimitReached:
+    return CONNECTION_LIMIT_REACHED;
+  case ProxyStatusError::TlsProtocolError:
+    return TLS_PROTOCOL_ERORR;
+  case ProxyStatusError::TlsCertificateError:
+    return TLS_CERTIFICATE_ERORR;
+  case ProxyStatusError::TlsAlertReceived:
+    return TLS_ALERT_RECEIVED;
+  case ProxyStatusError::HttpRequestError:
+    return HTTP_REQUEST_ERROR;
+  case ProxyStatusError::HttpRequestDenied:
+    return HTTP_REQUEST_DENIED;
+  case ProxyStatusError::HttpResponseIncomplete:
+    return HTTP_RESPONSE_INCOMPLETE;
+  case ProxyStatusError::HttpResponseHeaderSectionSize:
+    return HTTP_RESPONSE_HEADER_SECTION_SIZE;
+  case ProxyStatusError::HttpResponseHeaderSize:
+    return HTTP_RESPONSE_HEADER_SIZE;
+  case ProxyStatusError::HttpResponseBodySize:
+    return HTTP_RESPONSE_BODY_SIZE;
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+    return HTTP_RESPONSE_TRAILER_SECTION_SIZE;
+  case ProxyStatusError::HttpResponseTrailerSize:
+    return HTTP_RESPONSE_TRAILER_SIZE;
+  case ProxyStatusError::HttpResponseTransferCoding:
+    return HTTP_RESPONSE_TRANSFER_CODING;
+  case ProxyStatusError::HttpResponseContentCoding:
+    return HTTP_RESPONSE_CONTENT_CODING;
+  case ProxyStatusError::HttpResponseTimeout:
+    return HTTP_RESPONSE_TIMEOUT;
+  case ProxyStatusError::HttpUpgradeFailed:
+    return HTTP_UPGRADE_FAILED;
+  case ProxyStatusError::HttpProtocolError:
+    return HTTP_PROTOCOL_ERROR;
+  case ProxyStatusError::ProxyInternalResponse:
+    return PROXY_INTERNAL_RESPONSE;
+  case ProxyStatusError::ProxyInternalError:
+    return PROXY_INTERNAL_ERROR;
+  case ProxyStatusError::ProxyConfigurationError:
+    return PROXY_CONFIGURATION_ERROR;
+  case ProxyStatusError::ProxyLoopDetected:
+    return PROXY_LOOP_DETECTED;
+  }
+  return "-";
+}
+
+const absl::optional<ProxyStatusError>
+ProxyStatusUtils::fromStreamInfo(const StreamInfo& stream_info) {
+  // NB: This mapping from Envoy-specific ResponseFlag enum to Proxy-Status
+  // error enum is lossy, since ResponseFlag is really a bitset of many
+  // ResponseFlag enums. Here, we search the list of all known ResponseFlag values in
+  // enum order, returning the first matching ProxyStatusError.
+  if (stream_info.hasResponseFlag(ResponseFlag::FailedLocalHealthCheck)) {
```

Comment:
Is there a map somewhere between a response-flag and the corresponding `ProxyStatusError` enum value?
If so, can you add a link in the comment above.

Reply:
There isn't a canonical mapping besides what I've sketched here. I figure we can continue to iterate on these values over time, since the mapping is inherently lossy, and the RFC is still being finalized. I also didn't think it would be useful to encode this mapping in a const CONSTRUCT_ON_FIRST_USE map and iterate through the ResponseFlag values, since there's no easy ordered iterator through those values.